File: Arrays101/#496 - Next Greater Element - Easy/Solution.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
#define MAX_NUM 10001

int* nextGreaterElement(int* nums1, int nums1Size, int* nums2, int nums2Size, int* returnSize) {
    int* result = (int*)malloc(sizeof(int) * nums1Size);
    int next_greater[MAX_NUM];  
    int stack[MAX_NUM];
    int top = -1;

    for (int i = 0; i < MAX_NUM; i++) {
        next_greater[i] = -1;
    }

    for (int i = 0; i < nums2Size; i++) {
        int current = nums2[i];
        while (top >= 0 && current > stack[top]) {
            int prev = stack[top--];
            next_greater[prev] = current;
        }
        stack[++top] = current;
    }

    for (int i = 0; i < nums1Size; i++) {
        result[i] = next_greater[nums1[i]];
    }

    *returnSize = nums1Size;
    return result;
}

File: Arrays101/#496 - Next Greater Element - Easy/Solution.py
class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        next_greater = {}
        stack = []

        for num in nums2:
            while stack and num > stack[-1]:
                prev = stack.pop()
                next_greater[prev] = num
            stack.append(num)

        for num in stack:
            next_greater[num] = -1

        return [next_greater[num] for num in nums1]